subsys/net/lib/ptp/transport.c

```diff
@@ -0,0 +1,354 @@
+/*
+ * Copyright (c) 2024 BayLibre SAS
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(ptp_transport, CONFIG_PTP_LOG_LEVEL);
+
+#include <zephyr/kernel.h>
+
+#include <zephyr/net/socket.h>
+
+#include "transport.h"
+
+#define INTERFACE_NAME_LEN (32)
+
+#define IP_MULTICAST_IP "224.0.1.129"
+#define IP6_MULTICAST_IP "FF0E:0:0:0:0:0:0:181"
+
+#if CONFIG_PTP_UDP_IPv4_PROTOCOL
+static struct in_addr mcast_addr;
+#elif CONFIG_PTP_UDP_IPv6_PROTOCOL
+static struct in6_addr mcast_addr;
+#else
+#error "Chosen PTP transport protocol not implemented"
+#endif
+
+static int transport_socket_open(struct net_if *iface, struct sockaddr *addr)
+{
+	static const int feature_on = 1;
+	static const uint8_t priority = NET_PRIORITY_CA;
+	static const uint8_t ts_mask = SOF_TIMESTAMPING_TX_HARDWARE | SOF_TIMESTAMPING_RX_HARDWARE;
+	struct ifreq ifreq = { 0 };
+	int socket = zsock_socket(addr->sa_family, SOCK_DGRAM, IPPROTO_UDP);
+
+	if (net_if_get_by_iface(iface) < 0) {
+		LOG_ERR("Failed to obtain interface index");
+		return -1;
+	}
+
+	if (socket < 0) {
+		return -1;
+	}
+
+	if (zsock_setsockopt(socket, SOL_SOCKET, SO_REUSEADDR, &feature_on, sizeof(feature_on))) {
+		LOG_ERR("Failed to set SO_REUSEADDR");
+		goto error;
+	}
+
+	if (zsock_bind(socket, addr, sizeof(*addr))) {
+		LOG_ERR("Failed to bind socket");
+		goto error;
+	}
+
+	int cnt = net_if_get_name(iface, ifreq.ifr_name, INTERFACE_NAME_LEN);
+
+	if (cnt > 0 && zsock_setsockopt(socket,
+					SOL_SOCKET,
+					SO_BINDTODEVICE,
+					ifreq.ifr_name,
+					sizeof(ifreq.ifr_name))) {
+		LOG_ERR("Failed to set socket binding to an interface");
+		goto error;
+	}
+
+	if (zsock_setsockopt(socket, SOL_SOCKET, SO_TIMESTAMPING, &ts_mask, sizeof(ts_mask))) {
+		LOG_ERR("Failed to set SO_TIMESTAMPING");
+		goto error;
+	}
+
+	if (zsock_setsockopt(socket, SOL_SOCKET, SO_PRIORITY, &priority, sizeof(priority))) {
+		LOG_ERR("Failed to set SO_PRIORITY");
+		goto error;
+	}
+
+	return socket;
+error:
+	zsock_close(socket);
+	return -1;
+}
+
+static int transport_join_multicast(struct ptp_port *port)
+{
+	if (IS_ENABLED(CONFIG_PTP_UDP_IPv4_PROTOCOL)) {
+		struct ip_mreqn mreqn = {0};
+
+		memcpy(&mreqn.imr_multiaddr, &mcast_addr, sizeof(struct in_addr));
+		mreqn.imr_ifindex = net_if_get_by_iface(port->iface);
+
+		zsock_setsockopt(port->socket[1], IPPROTO_IP,
+				 IP_ADD_MEMBERSHIP, &mreqn, sizeof(mreqn));
+	} else {
```

Comment:
This prevents IPv4 and IPv6 use at the same time, shouldn't we allow it?

Reply:
I need to investigate the whole procedure for joining a multicast group. As you see I use only one socket of an interface because  joining multicast group on the second socket returns error.